crates/ext-processor/src/service.rs

```diff
@@ -651,15 +985,124 @@
+            .unwrap();
+
+        info!(
+            message = "combine decision",
+            accept = decision.accept,
+            restrict = decision.restrict,
+            unknown = decision.unknown,
+            score = decision.pignistic().restrict,
+            outcome = outcome.to_string(),
+            observe_only = thresholds.observe_only,
+            // array values aren't handled well unfortunately, coercing to comma-separated values seems to be the best option
+            tags = decision_components
+                .tags
+                .iter()
+                .map(|s| s.as_str())
+                .collect::<Vec<&str>>()
+                .to_vec()
+                .join(","),
+        );
+
+        match outcome {
+            bulwark_wasm_sdk::Outcome::Trusted
+            | bulwark_wasm_sdk::Outcome::Accepted
+            // suspected requests are monitored but not rejected
+            | bulwark_wasm_sdk::Outcome::Suspected => {
+                let result = Self::allow_request_body(&sender).await;
+                // TODO: must perform proper error handling on sender results, sending can fail
+                if let Err(err) = result {
+                    debug!(message = format!("send error: {}", err));
+                }
+            },
+            bulwark_wasm_sdk::Outcome::Restricted => {
+                if !thresholds.observe_only {
+                    info!(message = "process response", status = 403);
+                    let result = Self::block_request_body(&sender).await;
+                    // TODO: must perform proper error handling on sender results, sending can fail
+                    if let Err(err) = result {
+                        debug!(message = format!("send error: {}", err));
+                    }
+
+                    // Normally we initiate feedback after the response phase, but if we skip the response phase
+                    // we need to do it here instead.
+                    Self::handle_decision_feedback(
+                        decision_components,
+                        outcome,
+                        plugin_instances,
+                        timeout_duration,
+                    ).await;
+                    // Short-circuit if restricted, we can skip the response phase
+                    return;
+                } else {
+                    // Don't receive a body when we would have otherwise blocked if we weren't in monitor-only mode
+                    let result = Self::allow_request_body(&sender).await;
+                    // TODO: must perform proper error handling on sender results, sending can fail
+                    if let Err(err) = result {
+                        debug!(message = format!("send error: {}", err));
+                    }
+                }
+            },
+        }
+
+        if let Ok(http_resp) = Self::prepare_response(&mut stream, http_req.version()).await {
```

Comment:
Agree that switching to early return would be an improvement, but right now these functions aren't really doing error handling just yet and introducing that is better left for an explicit refactor since a decent number of function signatures would need to change in order to accommodate it.